openproblems/tasks/spatial_decomposition/methods/cell2location.py

```diff
@@ -0,0 +1,249 @@
+        max_epochs_sp = max_epochs_sp or 30000
+        num_samples = num_samples or 1000
+
+    adata_sc, adata = split_sc_and_sp(adata)
+
+    if "tech" in adata_sc.obs.columns:
+        adata_sc.obs["batch_key"] = adata_sc.obs["tech"].copy()
+    else:
+        adata_sc.obs["batch_key"] = "all"
+
+    # REFERENCE SIGNATURE ESTIMATION FROM SCRNA
+    # prepare anndata for the regression model
+    RegressionModel.setup_anndata(
+        adata=adata_sc,
+        # 10X reaction / sample / batch
+        batch_key="batch_key",
+        # cell type, covariate used for constructing signatures
+        labels_key="label",
+    )
+    sc_model = RegressionModel(adata_sc)
+    sc_model.train(max_epochs=max_epochs_sc, batch_size=sc_batch_size)
+    # In this section, we export the estimated cell abundance
+    # (summary of the posterior distribution).
+    adata_sc = sc_model.export_posterior(
+        adata_sc,
+        sample_kwargs={"num_samples": num_samples, "batch_size": sc_batch_size},
+    )
+    # export estimated expression in each cluster
+    if "means_per_cluster_mu_fg" in adata_sc.varm.keys():
+        inf_aver = adata_sc.varm["means_per_cluster_mu_fg"][
+            [
+                f"means_per_cluster_mu_fg_{i}"
+                for i in adata_sc.uns["mod"]["factor_names"]
+            ]
+        ].copy()
+    else:
+        inf_aver = adata_sc.var[
+            [
+                f"means_per_cluster_mu_fg_{i}"
+                for i in adata_sc.uns["mod"]["factor_names"]
+            ]
+        ].copy()
+    inf_aver.columns = adata_sc.uns["mod"]["factor_names"]
+
+    # SPATIAL MAPPING
+    # find shared genes and subset both anndata and reference signatures
+    intersect = np.intersect1d(adata.var_names, inf_aver.index)
+    adata = adata[:, intersect].copy()
+    inf_aver = inf_aver.loc[intersect, :].copy()
+
+    # prepare anndata for cell2location model
+    adata.obs["sample"] = "all"
+    Cell2location.setup_anndata(adata=adata, batch_key="sample")
+    if not amortised:
+        # create and train the model
+        st_model = Cell2location(
+            adata,
+            cell_state_df=inf_aver,
+            # the expected average cell abundance: tissue-dependent
+            # hyper-prior which can be estimated from paired histology:
+            # here = average in the simulated dataset
+            N_cells_per_location=N_cells_per_location,
+            # hyperparameter controlling normalisation of
+            # within-experiment variation in RNA detection:
+            detection_alpha=detection_alpha,
+        )
+
+        st_model.train(
+            max_epochs=max_epochs_sp,
+            # train using full data (batch_size=None)
+            batch_size=st_batch_size,
+            # use all data points in training because
+            # we need to estimate cell abundance at all locations
+            train_size=1,
+        )
+    else:
+        # create and train the model
+        st_model = Cell2location(
```

Comment:
What if instead of calling this in both arms of the if statement, we just create
```
cell2location_kwargs = dict(
            cell_state_df=inf_aver,
            # the expected average cell abundance: tissue-dependent
            # hyper-prior which can be estimated from paired histology:
            # here = average in the simulated dataset
            N_cells_per_location=N_cells_per_location,
            # hyperparameter controlling normalisation of
            # within-experiment variation in RNA detection:
            detection_alpha=detection_alpha,
        )
if amortised:
    cell2location_kwargs['amortised'] = True
    [...]

st_model = Cell2location(**cell2location_kwargs)
```